hy/compiler.py

```diff
@@ -414,33 +391,21 @@ def imports_as_stmts(self, expr):
         self.imports = defaultdict(set)
         return ret.stmts
 
-    def compile_atom(self, atom_type, atom):
-        if isinstance(atom_type, string_types):
-            atom_type = ast_str(atom_type)
-        if atom_type in _compile_table:
-            # _compile_table[atom_type] is a method for compiling this
-            # type of atom, so call it. If it has an extra parameter,
-            # pass in `atom_type`.
-            atom_compiler = _compile_table[atom_type]
-            arity = hy.inspect.get_arity(atom_compiler)
-            # Compliation methods may mutate the atom, so copy it first.
-            atom = copy.copy(atom)
-            ret = (atom_compiler(self, atom, atom_type)
-                   if arity == 3
-                   else atom_compiler(self, atom))
-            if not isinstance(ret, Result):
-                ret = Result() + ret
-            return ret
+    def compile_atom(self, atom):
         if not isinstance(atom, HyObject):
             atom = wrap_value(atom)
-            if isinstance(atom, HyObject):
-                spoof_positions(atom)
-                return self.compile_atom(type(atom), atom)
+            if not isinstance(atom, HyObject):
+                return
+            spoof_positions(atom)
+        if type(atom) not in _model_compilers:
+            return
+        # Compilation methods may mutate the atom, so copy it first.
+        atom = copy.copy(atom)
```

Comment:
Not necessarily something that needs to be done now, but what would it require to make atoms immutable? I sort of thought that would be the case already.

Reply:
That's #1542. Basically we'll just need to remove the remaining mutating code from the compiler.